pkg/reconciler/revision/background.go

```diff
@@ -0,0 +1,260 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package revision
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"sync"
+	"time"
+
+	"github.com/google/go-containerregistry/pkg/authn/k8schain"
+	"go.uber.org/zap"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/util/workqueue"
+	v1 "knative.dev/serving/pkg/apis/serving/v1"
+)
+
+// imageResolver is an interface used mostly to mock digestResolver for tests.
+type imageResolver interface {
+	Resolve(ctx context.Context, image string, opt k8schain.Options, registriesToSkip sets.String) (string, error)
+}
+
+// backgroundResolver performs background downloads of image digests.
+type backgroundResolver struct {
+	logger *zap.SugaredLogger
+
+	resolver imageResolver
+	enqueue  func(types.NamespacedName)
+
+	queue workqueue.Interface
+
+	mu      sync.Mutex
+	results map[types.NamespacedName]*resolveResult
+}
+
+// resolveResult is the overall result for a particular revision. We create a
+// workItem for each container we need to resolve for the overall result.
+type resolveResult struct {
+	statuses  []v1.ContainerStatus
+	err       error
+	remaining int
+
+	opt              k8schain.Options
+	registriesToSkip sets.String
+
+	completionCallback func()
+}
+
+// workItem is a single task submitted to the queue, to resolve a single image
+// for a resolveResult.
+type workItem struct {
+	result  *resolveResult
+	timeout time.Duration
+
+	name  string
+	image string
+	index int
+}
+
+func newBackgroundResolver(logger *zap.SugaredLogger, resolver imageResolver, enqueue func(types.NamespacedName)) *backgroundResolver {
+	r := &backgroundResolver{
+		logger: logger,
+
+		resolver: resolver,
+		enqueue:  enqueue,
+
+		results: make(map[types.NamespacedName]*resolveResult),
+		queue:   workqueue.NewNamed("digests"),
+	}
+
+	return r
+}
+
+// Start starts the worker threads and runs maxInFlight workers until the stop
+// channel is closed. It returns a done channel which will be closed when all
+// workers have exited.
+func (r *backgroundResolver) Start(stop <-chan struct{}, maxInFlight int) (done chan struct{}) {
+	var wg sync.WaitGroup
+
+	// Run the worker threads.
+	wg.Add(maxInFlight)
+	for i := 0; i < maxInFlight; i++ {
+		go func() {
+			defer wg.Done()
+			for {
+				item, shutdown := r.queue.Get()
+				if shutdown {
+					return
+				}
+
+				rrItem, ok := item.(*workItem)
+				if !ok {
+					r.logger.Fatalf("Unexpected work item type: want: %s, got: %s\n",
+						reflect.TypeOf(&workItem{}).Name(), reflect.TypeOf(item).Name())
+				}
```

Comment:
just use `%T`

Reply:
good point, copied this from networking prober, but will fix